### PR TITLE
VerrazzanoHelidon CRD and Controller

### PR DESCRIPTION
**Pull request by arjav-desai**
# Description

This is change contains CRD and workload related integration along with initial implementation of controller for Helidon applications.

# How has this been tested?

Provide details about how you tested this changed, if necessary
provide instructions to reproduce the testing.

- [x] Tested locally in my own environment
- [ ] Successful acceptance test run on CI server
- [ ] Other (specify)

# Checklist 

As the author of this PR, I have:

- [x] Checked my code follows the style guidelines, ran `go fmt`
- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated godoc for all public functions
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate
- [x] Checked that I correctly spelled trademark and product names
- [x] Used inclusive language and neutral tone
- [x] Not included any sensitive information or hardcoded credentials

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
+// Copyright (c) 2021, Oracle and/or its affiliates.
+// Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+package helidonworkload
+
+import (
```
Comment (kminder) [inline application-operator/controllers/helidonworkload/helidonworkload_controller.go:6]: I believe our convention is to not mix the standard packages with external packages.

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+		if k8serrors.IsNotFound(err) {
+			log.Info("VerrazzanoHelidonWorkload has been deleted", "name", req.NamespacedName)
+		} else {
+			log.Error(err, "Failed to fetch VerrazzanoHelidonWorkload", "name", req.NamespacedName)
+		}
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+	log.Info("Got the workload", "apiVersion", workload.APIVersion, "kind", workload.Kind)
+
+	//TODO: Find the resource object to record the event to, default is the parent appConfig - done in OAM
```
Comment (kminder) [inline application-operator/controllers/helidonworkload/helidonworkload_controller.go:72]: Should the TODO be removed or done?

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+	var workload vzapi.VerrazzanoHelidonWorkload
+	if err := r.Get(ctx, req.NamespacedName, &workload); err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Info("VerrazzanoHelidonWorkload has been deleted", "name", req.NamespacedName)
+		} else {
+			log.Error(err, "Failed to fetch VerrazzanoHelidonWorkload", "name", req.NamespacedName)
+		}
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+	log.Info("Got the workload", "apiVersion", workload.APIVersion, "kind", workload.Kind)
```
Comment (kminder) [inline application-operator/controllers/helidonworkload/helidonworkload_controller.go:70]: I'm not a fan of the "Got" in the message.  I'd actually prefer this message be before the Get.

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+	//TODO: Find the resource object to record the event to, default is the parent appConfig - done in OAM
+
+	//unwrap the apps/DeploymentSpec and meta/ObjectMeta
+	deploy, err := r.convertWorkloadToDeployment(&workload)
+	if err != nil {
+		log.Error(err, "Failed to convert workload to deployment")
+		//TODO: OAM is doing Wait and formatting error
```
Comment (kminder) [inline application-operator/controllers/helidonworkload/helidonworkload_controller.go:78]: TODO

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+	d.ObjectMeta.SetAnnotations(workload.Spec.DeploymentTemplate.Metadata.GetAnnotations())
+	// Set replication deployment from workload spec
+	d.Spec.Replicas = workload.Spec.DeploymentTemplate.Replicas
+	// Set PodSpec on deployment's PodTemplate from workload spec
+	workload.Spec.DeploymentTemplate.PodSpec.DeepCopyInto(&d.Spec.Template.Spec)
+	//making sure pods have same label as selector on deployment
+	d.Spec.Template.ObjectMeta.SetLabels(map[string]string{
+		labelKey: string(workload.GetUID()),
+	})
+	//k8s server-side patch complains if the protocol is not set on container port
```
Comment (kminder) [inline application-operator/controllers/helidonworkload/helidonworkload_controller.go:178]: Is this comment more confusing now than helpful?

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+					Name:       deploy.GetName(),
+					Port:       deploy.Spec.Template.Spec.Containers[0].Ports[0].ContainerPort,
+					TargetPort: intstr.FromInt(int(deploy.Spec.Template.Spec.Containers[0].Ports[0].ContainerPort)),
+					Protocol:   corev1.ProtocolTCP,
+				},
+			}
+		}
+		if y, err := yaml.Marshal(s); err != nil {
+			r.Log.Error(err, "Failed to convert service to yaml")
+			r.Log.Info("Set service in json ", "ServiceJson", s)
```
Comment (kminder) [inline application-operator/controllers/helidonworkload/helidonworkload_controller.go:236]: I'm unclear what "Set" means in these messages.

File: application-operator/controllers/helidonworkload/helidonworkload_controller_test.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	mocker.Finish()
+	assert.NoError(err)
+}
+
+// TestReconcileWorkloadNotFound tests reconciling a VerrazzanoHelidonWorkload when the workload
+// cannot be fetched. This happens when the workload has been deleted by the OAM runtime.
+// GIVEN a VerrazzanoHelidonWorkload resource has been deleted
+// WHEN the controller Reconcile function is called and we attempt to fetch the workload
+// THEN return success from the controller as there is nothing more to do
+func TestReconcileWorkloadNotFound(t *testing.T) {
```
Comment (kminder) [inline application-operator/controllers/helidonworkload/helidonworkload_controller_test.go:61]: Not many tests here.  What is the unit test coverage.  We have an 80% coverage policy.

File: examples/hello-helidon/hello-helidon-comp.yaml [hunk 8]
```diff
@@ -20,4 +20,4 @@ spec:
         image: "ghcr.io/verrazzano/example-helidon-greet-app-v1:0.1.10-3-20201016220428-56fb4d4"
         ports:
           - containerPort: 8080
-            name: http
+            name: http
```
Comment (kminder) [inline examples/hello-helidon/hello-helidon-comp.yaml:23]: Put back the newline.

File: platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_verrazzanohelidonworkloads.yaml [hunk 9]
```diff
@@ -0,0 +1,6005 @@
...
+    kind: VerrazzanoHelidonWorkload
+    listKind: VerrazzanoHelidonWorkloadList
+    plural: verrazzanohelidonworkloads
+    singular: verrazzanohelidonworkload
+  scope: Namespaced
+  versions:
+    - name: v1alpha1
+      schema:
+        openAPIV3Schema:
+          description: VerrazzanoHelidonWorkload is the Schema for the verrazzanohelidonworkloads
```
Comment (kminder) [inline platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_verrazzanohelidonworkloads.yaml:24]: remove "the" in the "for the verrazzanohelidonworkloads".

File: platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_vzhelidon.yaml [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+metadata:
+  name: verrazzanohelidonworkloads.oam.verrazzano.io
+spec:
+  childResourceKinds:
+    - apiVersion: v1
+      kind: Deployment
+    - apiVersion: v1
+      kind: Service
+  definitionRef:
+    name: verrazzanohelidonworkloads.oam.verrazzano.io
```
Comment (kminder) [inline platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_vzhelidon.yaml:15]: Newline at end.

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go (hunk 3 above)
Comment (arjav-desai) [reply application-operator/controllers/helidonworkload/helidonworkload_controller.go:70]: Removing "Got" but can't move it before Get as workload would be empty then.

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go (hunk 1 above)
Comment (arjav-desai) [reply application-operator/controllers/helidonworkload/helidonworkload_controller.go:6]: Sorted the imports.

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go (hunk 5 above)
Comment (arjav-desai) [reply application-operator/controllers/helidonworkload/helidonworkload_controller.go:178]: Removed it.

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go (hunk 6 above)
Comment (arjav-desai) [reply application-operator/controllers/helidonworkload/helidonworkload_controller.go:236]: Clarified it.

File: platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_vzhelidon.yaml (hunk 10 above)
Comment (arjav-desai) [reply platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_vzhelidon.yaml:15]: Added.

File: examples/hello-helidon/hello-helidon-comp.yaml (hunk 8 above)
Comment (arjav-desai) [reply examples/hello-helidon/hello-helidon-comp.yaml:23]: Added.

File: platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_verrazzanohelidonworkloads.yaml (hunk 9 above)
Comment (arjav-desai) [reply platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_verrazzanohelidonworkloads.yaml:24]: Fixed it.

**Review by fredtibbitts — COMMENTED** [review]
I don't have any substantive feedback beyond what Kevin has already noted. Definitely agree on more unit tests.

One minor nit, your comments seem inconsistent. Some start with a capital letter, some lowercase, some have a space after the `//`, some do not. Not a big deal.

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go (hunk 2 above)
Comment (arjav-desai) [reply application-operator/controllers/helidonworkload/helidonworkload_controller.go]: Removed.

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go (hunk 4 above)
Comment (arjav-desai) [reply application-operator/controllers/helidonworkload/helidonworkload_controller.go]: Removed.

File: application-operator/controllers/helidonworkload/helidonworkload_controller_test.go (hunk 7 above)
Comment (arjav-desai) [reply application-operator/controllers/helidonworkload/helidonworkload_controller_test.go:66]: Added tests and we are at 77% with it with major uncovered lines being error catch on translate and apply patch.

**Review by kminder — APPROVED** [review]
Approved.

**Pull request merged**